pupil-spa/src/app/services/user/user.service.spec.ts

```diff
@@ -1,87 +1,72 @@
-import { MockBackend } from '@angular/http/testing';
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
+import { HttpClient } from '@angular/common/http';
 import { UserService } from './user.service';
 import { StorageService } from '../storage/storage.service';
 import * as mockLoginResponseBody from '../../login.userService.response.mock.json';
-import { AppConfigService, loadConfigMockService } from '../config/config.service';
+import { APP_CONFIG, AppConfigService, loadConfigMockService } from '../config/config.service';
 
 const shouldNotExecute = () => {
   expect('this code').toBe('not executed');
 };
 
 let userService: UserService;
 let storageService: StorageService;
-let mockBackend: MockBackend;
 const pupilDataKey = 'pupil';
 const questionsDataKey = 'questions';
 const configDataKey = 'config';
 
 describe('UserService', () => {
+  let httpClient: HttpClient;
+  let httpTestingController: HttpTestingController;
 
   beforeEach(() => {
 
     const inject = TestBed.configureTestingModule({
-      imports: [HttpModule],
+      imports: [HttpClientTestingModule],
       providers: [
         UserService,
-        { provide: XHRBackend, useClass: MockBackend },
         StorageService
       ]
     });
 
+    // Inject the http service and test controller for each test
+    httpClient = TestBed.get(HttpClient);
+    httpTestingController = TestBed.get(HttpTestingController);
+
     userService = inject.get(UserService);
     storageService = inject.get(StorageService);
-    mockBackend = inject.get(XHRBackend);
-
   });
 
   describe('login', () => {
-
     it('should persist response body to storage', () => {
-
-      mockBackend.connections.subscribe((connection) => {
-        connection.mockRespond(new Response(new ResponseOptions({
-          body: JSON.stringify(mockLoginResponseBody),
-          status: 200
-        })));
-      });
-
       spyOn(storageService, 'setItem');
 
       userService.login('abc12345', '9999a').then(() => {
-        expect(storageService.setItem)
-          .toHaveBeenCalledWith(pupilDataKey, mockLoginResponseBody[pupilDataKey]);
-        expect(storageService.setItem)
-          .toHaveBeenCalledWith(questionsDataKey, mockLoginResponseBody[questionsDataKey]);
-        expect(storageService.setItem)
-        .toHaveBeenCalledWith(configDataKey, mockLoginResponseBody[configDataKey]);
-      },
+          expect(storageService.setItem)
+            .toHaveBeenCalledWith(pupilDataKey, mockLoginResponseBody[pupilDataKey]);
+          expect(storageService.setItem)
+            .toHaveBeenCalledWith(questionsDataKey, mockLoginResponseBody[questionsDataKey]);
+          expect(storageService.setItem)
+            .toHaveBeenCalledWith(configDataKey, mockLoginResponseBody[configDataKey]);
+        },
         (err) => {
           shouldNotExecute();
         });
+
+      const req = httpTestingController.expectOne(`${APP_CONFIG.authURL}`);
```

Comment:
What's the reason behind adding the variable in a string template?
does it not work with just `expectOne(APP_CONFIG.authURL)` ?

Reply:
That was the existing pattern so I kept it consistent.  I suppose the property could be set to a non-string value?

Reply:
I would expect not but it is indeed safeguard if that scenario can potentially happen.